File: leetcode/two pointer/76. Minimum Window Substring/js/main.js
// For most substring problem, we are given a string and need to find a substring of it which satisfy some restrictions.
// A general way is to use a hashmap assisted with two pointers.

// Time O(N)
// Space O(N)
const minWindow = (s, t) => {
  let map = {};

  for (let x of t) {
    map[x] = ~~map[x] + 1;
  }

  let start = 0;
  let end = 0;
  let INF = Number.MAX_VALUE;
  let k = Object.keys(map).length;
  let minLen = INF;
  let minIndex = INF;

  while (end < s.length) {
    if (map[s[end]] == 1) {
      k--;
    }

    map[s[end++]]--;

    while (k == 0) {
      // update minLength here if finding minimum
      if (minLen > end - start) {
        minLen = end - start;
        minIndex = start;
      }

      if (map[s[start]] == 0) {
        k++;
      }

      map[s[start++]]++;
    }
    // update maxLength here if finding maximum
  }

  return minLen == INF ? '' : s.substring(minIndex, minLen + minIndex);
};

File: leetcode/two pointer/76. Minimum Window Substring/py/main.py
import collections
import math


class Solution:
    # Time O(N)
    # Space O(N)
    def minWindow(self, s, t):
        count = collections.Counter(t)

        start, end = 0, 0
        k = len(count)
        minIndex = math.inf
        minLen = math.inf

        while end < len(s):
            if count[s[end]] == 1:
                k -= 1

            count[s[end]] -= 1
            end += 1

            while k == 0:
                if minLen > end - start:
                    minLen = end - start
                    minIndex = start

                if count[s[start]] == 0:
                    k += 1

                count[s[start]] += 1
                start += 1

        return '' if minLen == math.inf else s[minIndex:minIndex + minLen]
